### PR TITLE
Refreshtoken lifetime + 0.3.0 version bump

### PR DESCRIPTION
**Pull request by DeepDiver1975**
Follow up on #174

### CODE REVIEW

**Comment by PVince81** [comment]
unit tests ?

File: lib/Db/RefreshToken.php [hunk 1]
```diff
@@ -32,16 +32,36 @@
...
+		$this->setExpires(\time() + (self::EXPIRATION_DAYS * 86400));
+	}
+
+	/**
+	* Determines if an refresh token has expired.
+	*
+	* @return boolean true if the refresh token has expired, false otherwise.
+	*/
+	public function hasExpired() {
+		return \time() >= $this->getExpires();
```
Comment (PVince81) [inline lib/Db/RefreshToken.php]: use the TimeFactory everywhere so you can mock it in tests

File: lib/BackgroundJob/CleanUp.php [hunk 2]
```diff
@@ -22,28 +22,34 @@
...
 	public function __construct(
 		AuthorizationCodeMapper $authorizationCodeMapper,
-		AccessTokenMapper $accessTokenMapper) {
+		AccessTokenMapper $accessTokenMapper,
+		RefreshTokenMapper $refreshTokenMapper) {
 		$this->authorizationCodeMapper = $authorizationCodeMapper;
 		$this->accessTokenMapper = $accessTokenMapper;
+		$this->refreshTokenMapper = $refreshTokenMapper;
+
+		$this->setInterval(86400);
```
Comment (mrow4a) [inline lib/BackgroundJob/CleanUp.php]: magic number?

File: lib/Db/RefreshToken.php [hunk 3]
```diff
@@ -32,16 +32,36 @@
...
 		$this->addType('client_id', 'int');
 		$this->addType('user_id', 'string');
 		$this->addType('access_token_id', 'int');
+		$this->addType('expires', 'int');
+	}
+
+	/**
+	* Resets the expiry time to EXPIRATION_DAYS days from now.
+	*/
+	public function resetExpires() {
```
Comment (mrow4a) [inline lib/Db/RefreshToken.php]: Formating of comment is strange

File: lib/Controller/OAuthApiController.php [hunk 4]
```diff
@@ -206,6 +206,7 @@ public function generateToken($grant_type, $code = null,
 		$refreshToken->setClientId($client->getId());
 		$refreshToken->setUserId($userId);
 		$refreshToken->setAccessTokenId($accessToken->getId());
+		$refreshToken->resetExpires();
```
Comment (mrow4a) [inline lib/Controller/OAuthApiController.php]: What is the rationale behind allowing RefreshToken setting expiration, and not Controller?

**Comment by T0mWz** [comment]
Argg, applied this on our production environment & now run into a problem here..

When your AccessToken is expired and you retrieving a new one, your RefreshToken is regenerated where the AccessToken is linked too.

So, the lifetime of the RefreshToken is automatically extended with the duration of the AccessToken.
Have to update the existing RefreshToken with the ID of the new AccessToken, in stead of generate a new RefreshToken..

**Comment by T0mWz** [comment]
Ok, have fixed now. There are slightly some more adjustments. 
In "oauth2/lib/Db/RefreshTokenMapper.php" we need an extra function;
```
...
        /**
        * Update valid refresh token with a new Access Token
        */
        public function UpdateRefreshToken($AccessTokenId, $tokenId) {
                $this->logger->info('Update Refresh Token with new AccessToken id.', ['app' => $this->appName]);

                $sql = 'UPDATE `' . $this->tableName . '` SET `access_token_id`= ? WHERE `token` = ?';
                $stmt = $this->execute($sql, [$AccessTokenId, $tokenId], null, null);
                $stmt->closeCursor();
                return true;
        }
```

In "oauth2/lib/Controller/OAuthApiController.php" we will update the token when it's still valid & return the JSON.
```
@@ -169,8 +169,10 @@
 				try {
 					/** @var RefreshToken $refreshToken */
 					$refreshToken = $this->refreshTokenMapper->findByToken($refresh_token);
+					$this->logger->info('Refresh token found for client "' . $client->getName() . '".', ['app' => $this->appName]);
 				} catch (DoesNotExistException $exception) {
 					\OC::$server->getLogger()->logException($exception, ['app'=>__CLASS__]);
+					$this->logger->info('Refresh token not found for client "' . $client->getName() . '", it may have expired.', ['app' => $this->appName]);
 					return new JSONResponse(['error' => 'invalid_grant'], $statusCode);
 				}

@@ -179,14 +181,33 @@
 					return new JSONResponse(['error' => 'invalid_grant'], $statusCode);
 				}

-				$this->logger->info('A refresh token has been used by the client "' . $client->getName() . '" to request an access token.', ['app' => $this->appName]);
-
 				$userId = $refreshToken->getUserId();
 				$relatedAccessToken = new AccessToken();
-				$relatedAccessToken->setId($refreshToken->getAccessTokenId());
+				$relatedAccessToken->setId($refreshToken->getAccessTokenId());
 				$this->accessTokenMapper->delete($relatedAccessToken);
-				$this->refreshTokenMapper->delete($refreshToken);
-				break;
+
+        		        $token = Utilities::generateRandom();
+        		        $accessToken = new AccessToken();
+		                $accessToken->setToken($token);
+                		$accessToken->setClientId($client->getId());
+                		$accessToken->setUserId($userId);
+        		        $accessToken->resetExpires();
+		                $this->accessTokenMapper->insert($accessToken);
+				$refreshToken = $this->refreshTokenMapper->UpdateRefreshToken($accessToken->getId(), $refresh_token);
+
+				$this->logger->info('Current refresh token for user "' . $userId . '" with client "' . $client->getName() . '" is still valid and has been used to request an access token.', ['app' => $this->appName]);
+
+		                return new JSONResponse(
+	                        	[
+	                        	        'access_token' => $accessToken->getToken(),
+	                        	        'token_type' => 'Bearer',
+	                        	        'expires_in' => AccessToken::EXPIRATION_TIME,
+        	                	        'refresh_token' => $refresh_token,
+        	                	        'user_id' => $userId,
+        	                	        'message_url' => $this->urlGenerator->linkToRouteAbsolute($this->appName . '.page.authorizationSuccessful')
+        	        	        ]
+        		        );
+				//break;
 			default:
 				\OC::$server->getLogger()->debug("unhandled grant type: {$grant_type}", ['app'=>__CLASS__]);
 				return new JSONResponse(['error' => 'invalid_grant'], Http::STATUS_BAD_REQUEST);
@@ -206,6 +227,7 @@
 		$refreshToken->setClientId($client->getId());
 		$refreshToken->setUserId($userId);
 		$refreshToken->setAccessTokenId($accessToken->getId());
+		$refreshToken->resetExpires();
 		$this->refreshTokenMapper->insert($refreshToken);

 		return new JSONResponse(
``` 

Now I got only a nasty Exception when there isn't a RefreshToken found in the follow try statement for "oauth2/lib/Controller/OAuthApiController.php";
```
                                try {
                                        /** @var RefreshToken $refreshToken */
                                        $refreshToken = $this->refreshTokenMapper->findByToken($refresh_token);
                                        $this->logger->info('Refresh token found for client "' . $client->getName() . '".', ['app' => $this->appName]);
                                }
```

Maybe you can have  a look how this can be fixed in a more cleaner way.
```
ownCloud[12407]: {OCA\OAuth2\Controller\OAuthApiController} Exception: {"Exception":"OCP\\AppFramework\\Db\\DoesNotExistException","Message":"Did expect one result but found none when executing: query \"SELECT * FROM `*PREFIX*oauth2_refresh_tokens` WHERE `token` = ? AND expires > unix_timestamp(now())\"; parameters Array\n(\n    [0] => WzWrGCJWTJGBXJCrUK99s4txyDWZQO3iBlpMLaHsZu7NYyP87zisXvmlt9MHIwDD\n)\n; limit \"\"; offset \"\"","Code":0,"Trace":"#0 \/var\/www\/surfdrive-tst.1009\/lib\/public\/AppFramework\/Db\/Mapper.php(379): OCP\\AppFramework\\Db\\Mapper->findOneQuery('SELECT * FROM `...', Array, NULL, NULL)\n#1 \/var\/www\/surfdrive-tst.1009\/apps\/oauth2\/lib\/Db\/RefreshTokenMapper.php(84): OCP\\AppFramework\\Db\\Mapper->findEntity('SELECT * FROM `...', Array, NULL, NULL)\n#2 \/var\/www\/surfdrive-tst.1009\/apps\/oauth2\/lib\/Controller\/OAuthApiController.php(171): OCA\\OAuth2\\Db\\RefreshTokenMapper->findByToken('WzWrGCJWTJGBXJC...')\n#3 [internal function]: OCA\\OAuth2\\Controller\\OAuthApiController->generateToken(*** sensitive parameters replaced ***)\n#4 \/var\/www\/surfdrive-tst.1009\/lib\/private\/AppFramework\/Http\/Dispatcher.php(159): call_user_func_array(Array, Array)\n#5 \/var\/www\/surfdrive-tst.1009\/lib\/private\/AppFramework\/Http\/Dispatcher.php(89): OC\\AppFramework\\Http\\Dispatcher->executeController(Object(OCA\\OAuth2\\Controller\\OAuthApiController), 'generateToken')\n#6 \/var\/www\/surfdrive-tst.1009\/lib\/private\/AppFramework\/App.php(103): OC\\AppFramework\\Http\\Dispatcher->dispatch(Object(OCA\\OAuth2\\Controller\\OAuthApiController), 'generateToken')\n#7 \/var\/www\/surfdrive-tst.1009\/lib\/private\/AppFramework\/Routing\/RouteActionHandler.php(46): OC\\AppFramework\\App::main('OCA\\\\OAuth2\\\\Cont...', 'generateToken', Object(OC\\AppFramework\\DependencyInjection\\DIContainer), Array)\n#8 [internal function]: OC\\AppFramework\\Routing\\RouteActionHandler->__invoke(Array)\n#9 \/var\/www\/surfdrive-tst.1009\/lib\/private\/Route\/Router.php(342): call_user_func(Object(OC\\AppFramework\\Routing\\RouteActionHandler), Array)\n#10 \/var\/www\/surfdrive-tst.1009\/lib\/base.php(919): OC\\Route\\Router->match('\/apps\/oauth2\/ap...')\n#11 \/var\/www\/surfdrive-tst.1009\/index.php(55): OC::handleRequest()\n#12 {main}","File":"\/var\/www\/surfdrive-tst.1009\/lib\/public\/AppFramework\/Db\/Mapper.php","Line":294} [cd2169b2-6f62-42a8-bdfd-afeb09e986a4]
```

**Comment by mmattel** [comment]
Ping, any update on this?

**Comment by T0mWz** [comment]
> Ping, any update on this?

This is already open for a quite long period.
We have still applied these changes in the latest oAuth2 app release at SURF.
Can provide you an up-to-date patch if desired.

**Comment by mmattel** [comment]
> Can provide you an up-to-date patch if desired

Yes please 👍

**Comment by phil-davis** [comment]
@micbar pinging you so you are aware of this PR that has been here since 2018.
(and has minor conflicts)

**Comment by micbar** [comment]
As of discussion in RC closing this, obsolete.

**Comment by micbar** [comment]
Needs reasoning

**Comment by T0mWz** [comment]
@micbar ; From a security point of view, you do not want an infinite session duration. Where we (@SURF) also use an external Identity Provider (Shibboleth / LDAP), instead of local accounts, 
we also want to check once in the X period whether the user can still log in via the own IdP. For example when they forget to block the user within ownCloud.

**Comment by DeepDiver1975** [comment]
Reasoning:
All is moving into the direct of OpenID Connect - globally and here at ownCloud.
The OAuth2 app is loosing more and more reasons to exist and therefore we will not add new features for now.

**Comment by michaelstingl** [comment]
Policies (like expiration) are managed by the IdP with OIDC. All ownCloud clients will respect those.

**Comment by T0mWz** [comment]
@DeepDiver1975 , to make the challenge a little bigger 😅.. Now we support multiple login methods, both local and Shibboleth accounts (via SSO). Could this still be supported when switching to OIDC? I guess for local account, only the sessions handling is then used, like now is done by oAuth2 or is the "Identity Service" part also there to authenticate to a local database or something?

**Comment by DeepDiver1975** [comment]
@T0mWz I'm pretty much aware of the challenges in our setup! We need to have a closer look and discussion on how to get you migrated one day.

**Pull request closed without merging**